pkg/capabilities/triggers/mercury_trigger.go

```diff
@@ -0,0 +1,219 @@
+		val, err := mercury.Codec{}.WrapMercuryTriggerEvent(triggerEvent)
+		if err != nil {
+			return err
+		}
+
+		// Create a new CapabilityResponse with the MercuryTriggerEvent
+		capabilityResponse := capabilities.CapabilityResponse{
+			Value: val,
+		}
+
+		// If the FeedId is in the feedIdsForTriggerId map, then send the event to the channel.
+
+		wfID := o.triggerIDToWorkflowID[triggerID]
+		ch, ok := o.chans[workflowID(wfID)]
+		if ok {
+			ch <- capabilityResponse
+		} else {
+			return fmt.Errorf("no registration for %s", triggerID)
+		}
+	}
+	return nil
+}
+
+func (o *MercuryTriggerService) RegisterTrigger(ctx context.Context, callback chan<- capabilities.CapabilityResponse, req capabilities.CapabilityRequest) error {
+	wid := req.Metadata.WorkflowID
+
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	o.chans[workflowID(wid)] = callback // TODO: if this workflowID is already in the map, throw an error?
+	// set feedIdsForTriggerId
+	triggerID := o.GetTriggerID(req) // TODO: This is manadatory, so throw an error if its not present or already exists
+	feedIDs := o.GetFeedIDs(req)     // TODO: what if feedIds is empty? should we throw an error or allow it?
+	o.feedIDsForTriggerID[triggerID] = feedIDs
+
+	o.triggerIDToWorkflowID[triggerID] = wid
+	return nil
+}
+
+func (o *MercuryTriggerService) UnregisterTrigger(ctx context.Context, req capabilities.CapabilityRequest) error {
+	wid := req.Metadata.WorkflowID
+
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	ch, ok := o.chans[workflowID(wid)]
+	if ok {
+		close(ch)
+	}
+
+	triggerID := o.GetTriggerID(req)
+
+	delete(o.chans, workflowID(wid))
+	delete(o.feedIDsForTriggerID, triggerID)
+	delete(o.triggerIDToWorkflowID, triggerID)
+
+	return nil
+}
+
+// Get array of feedIds from CapabilityRequest req
+func (o *MercuryTriggerService) GetFeedIDs(req capabilities.CapabilityRequest) []int64 {
+	feedIDs := make([]int64, 0)
+	// Unwrap the inputs which should return pair (map, nil) and then get the feedIds from the map
+	if inputs, err := req.Inputs.Unwrap(); err == nil {
+		if feeds, ok := inputs.(map[string]interface{})["feedIds"].([]int64); ok {
+			// Copy to feedIds
+			feedIDs = append(feedIDs, feeds...)
+		}
+	}
+	return feedIDs
+}
+
+// Get the triggerId from the CapabilityRequest req map
+func (o *MercuryTriggerService) GetTriggerID(req capabilities.CapabilityRequest) string {
+	var triggerID string
+	// Unwrap the inputs which should return pair (map, nil) and then get the triggerId from the map
+	if inputs, err := req.Inputs.Unwrap(); err == nil {
+		if id, ok := inputs.(map[string]interface{})["triggerId"].(string); ok {
+			triggerID = id
+		}
+	}
+	return triggerID
+}
+
+func sha256Hash(s string) string {
+	hash := sha256.New()
+	hash.Write([]byte(s))
+	return hex.EncodeToString(hash.Sum(nil))
+}
+
+func generateTriggerEventID(reports []mercury.FeedReport) string {
+	// Let's hash all the feedIds and timestamps together
+	s := ""
+	for _, report := range reports {
```

Comment:
We should apply a sort before concatting them all so we get deterministic event IDs. I'd imagine we'd want to sort by feedId then timestamp afterwards since we could have identical timestamps for two different reports under two different feed IDs

Reply:
Done!